clients/drcachesim/tools/opcode_mix.cpp

```diff
@@ -222,72 +213,27 @@ opcode_mix_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
     if (shard->filetype == OFFLINE_FILE_TYPE_DEFAULT) {
         shard->error = "No file type found in this shard";
         return false;
+    } else if (shard->decode_cache == nullptr) {
+        std::string err = init_decode_cache(shard, dcontext_.dcontext, shard->filetype);
+        if (err != "") {
+            shard->error = err;
+            return false;
+        }
     }
 
     ++shard->instr_count;
 
-    app_pc decode_pc;
-    const app_pc trace_pc = reinterpret_cast<app_pc>(memref.instr.addr);
-    if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->filetype)) {
-        // The trace has instruction encodings inside it.
-        decode_pc = const_cast<app_pc>(memref.instr.encoding);
-        if (memref.instr.encoding_is_new) {
-            // The code may have changed: invalidate the cache.
-            shard->worker->opcode_data_cache.erase(trace_pc);
-        }
-    } else {
-        // Legacy trace support where we need the binaries.
-        if (!module_mapper_) {
-            shard->error =
-                "Module file path is missing and trace has no embedded encodings";
-            return false;
-        }
-        if (trace_pc >= shard->last_trace_module_start &&
```

Comment:
These last_ quick cache vars were important for performance: they should be in the library shouldn't they?

Reply:
Ah, thanks for reminding about this. I remember considering whether module_mapper_t should own this optimization or the decode_cache_t. Thoughts?

Reply:
I see that find_mapped_trace_bounds is intended for that exactly. Maybe we didn't want to add that opt to module_mapper_t::find_mapped_trace_address itself and let the user of module_mapper_t do it?

In any case, added it to decode_cache_base_t